### ISSUE TITLE
async load

### ISSUE THREAD

**Issue by AnimatorJeroen** [MolecularMatters/psd_sdk#5]
Hi Molecular Matters team,

How would we set up async load for the psd layers.
I believe the SDK already loads each channel async, which is nice. But we would like to extract layer data in parallel as well.

In the sample, above the layerMaskselection for loop, it says: 
line 279: "// extract all layers one by one. this should be done in parallel for maximum efficiency."

We try to launch ExtractLayer() using std::future, but then there is a conflict with Reader.

Thanks in advance.

Kind regards,
Jeroen

**Comment by MolecularMatters** [comment]
> I believe the SDK already loads each channel async, which is nice.

No, it doesn't do that out-of-the-box. Concurrency and threading is best left to the user, because different users have different preferences regarding their choice of library, e.g. std::future & async, Intel TBB, etc.
However, the SDK is built in a way that only does minimal parsing up-front, while allowing the user to work on parsing different sections and layers concurrently.

> We try to launch ExtractLayer() using std::future, but then there is a conflict with Reader.

What do you mean by "conflict with Reader"?
Internally, all IO operations go through the File interface, which only offers async operations. See also the remark for the File interface in the code:
"Note that the interface only offers asynchronous read operations. The reason for this is that asynchronous reads
allow for parallelizing file accesses to the same file, while still being able to add synchronous reads as a wrapper on top."

Furthermore, APIs that actually parse or load data all use a SyncFileReader on the stack, which is nothing more than a convenience wrapper for keeping track of the current file position for sequential reading. However, these readers are **not** shared by threads, because each reader lives on the stack and only uses async APIs offered by the file interface.

What are the conflicts you are seeing?

**Comment by hedzerkoolstra** [comment]
Hi,

Thanks for your quick reply. I'm working with Jeroen on this project and reproduced our problem.

We run the for loop as shown in the sample except we replace:
`ExtractLayer(document, file, allocator, layer);`
with:
`layerFutures.emplace_back(std::async(std::launch::async, ExtractLayer, document, file, allocator, layer));`
layerFutures being:
`std::vector<std::future<void>> layerFutures;`

We are not passing in &file and &allocator because we changed the code a bit and they are both already pointers. It works fine when we just call ExtractLayer so I don't think that's the issue.

When we run the code it breaks here because m_file is a nullptr:
```
void SyncFileReader::Read(void* buffer, uint32_t count)
{
	// do an asynchronous read, wait until it's finished, and update the file position
	File::ReadOperation op = m_file->Read(buffer, count, m_position);
	m_file->WaitForRead(op);

	m_position += count;
}
```
From your answer I assume this has something to do with the SyncFileReader not being shared by different threads. If so, what do you think is the best way to work around this?

Thanks!

Hedzer

**Comment by MolecularMatters** [comment]
> From your answer I assume this has something to do with the SyncFileReader not being shared by different threads. If so, what do you think is the best way to work around this?

I think you misunderstood.
Having SyncFileReaders on the stack is a **good** thing, because it means you cannot possible have any race conditions between different file readers, as they are never shared.

> We are not passing in &file and &allocator because we changed the code a bit and they are both already pointers.

Without knowing how you changed the code, I would say here lies your problem.
If you look at SyncFileReader, you will see that it takes a File* in its constructor. If m_file is nullptr in your case, it means that somebody passed an invalid file to e.g. the ExtractLayer function.
You should investigate how m_file of the reader can ever be nullptr - it's never touched after construction, and it's passed to the SyncFileReader directly in ExtractLayer:

`SyncFileReader reader(file);`

I'm assuming you're passing an invalid file pointer to ExtractLayer.

**Comment by hedzerkoolstra** [comment]
Thanks for answering. I understand that this seems the most likely problem. But I ran some checks today and I'm pretty sure it isn't. 

I tried the sample code, without our own changes. I executed ExtractLayer with &file and &allocator and everything works fine. But when I pass the function into std::async (with &file and &allocator) it gives the same error.

I then put a breakpoint just before ReadFromFileBE within ExtractLayer and it recognizes m_file. But as soon as ReadFromFileBE executes it gives the error.

Should calling ExtractLayer with std::async not give any problems normally? Maybe I'm overlooking something.

Thanks!

**Comment by MolecularMatters** [comment]
> Should calling ExtractLayer with std::async not give any problems normally? Maybe I'm overlooking something.

No, this definitely does not cause any problems.

My guess would be that you're not paying attention to when and how those layers you try to extract asynchronously are actually used, probably having at least one race condition in your code.
std:.async is not some magic wand you can wave around to make your code multithreaded, you still need to make sure that data is only accessed when it is actually available.

As an example, this works without problems (but is not terribly efficient):

```
for (unsigned int i = 0; i < layerMaskSection->layerCount; ++i)
{
	Layer* layer = &layerMaskSection->layers[i];

	layerFutures.emplace_back(std::async(std::launch::async, ExtractLayer, document, &file, &allocator, layer));
}

for (unsigned int i = 0; i < layerMaskSection->layerCount; ++i)
{
	layerFutures[i].wait();
}
```

**Comment by hedzerkoolstra** [comment]
Thanks for your help guys, We've been trying to work it out. We've added the .wait() function as shown in our example and now get a different error: Malformed RLE data encountered.

As far as I understand the code this has something to do with rleDataSize not being valid somehow. Could it be that m_position is changing when a new thread is started and the old thread is not yet finished and therefor dataCount in ReadChannelDataRLE() is wrong?

**Comment by MolecularMatters** [comment]
> Could it be that m_position is changing when a new thread is started and the old thread is not yet finished and therefor dataCount in ReadChannelDataRLE() is wrong?

It seems to me you don't understand the basics of threading, local variables on the stack, etc.
Once again, ParseImageDataSection() uses a local SyncFileReader on the stack, so it does not matter which threads try to read the image data section from which file. Each SyncFileReader has it's own state and **cannot** be influenced by any other readers which live on the stack of any of the other threads.

I suggest you get a good understanding of multi-threaded programming first.
I cannot continue debugging and explaining these non-issues for you, and will therefore close this issue.

If you think the RLE decoder in imageUtil::DecompressRle() is broken, then feel free to open another issue for that.
